ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,344 @@
+      separateDeclinedAndStillInRace ::
+        [(FetchDecision (ChainSuffix header), peerInfo)] ->
+        WithDeclined peerInfo (Maybe (ChainSuffix header, [(ChainSuffix header, peerInfo)]))
+      separateDeclinedAndStillInRace decisions = do
+        inRace <-
+          catMaybes
+            <$> traverse
+              ( \(decision, peer) ->
+                  case decision of
+                    Left reason -> tell (List [(reason, peer)]) >> pure Nothing
+                    Right candidate -> pure $ Just (candidate, peer)
+              )
+              decisions
+        return $ ((,inRace) . fst . NE.head) <$> nonEmpty inRace
+
+-- | Given _the_ candidate fragment to sync from, and a list of peers (with
+-- their corresponding candidate fragments), choose which peer to sync _the_
+-- candidate fragment from.
+--
+-- We first filter out all the peers that cannot even serve a reasonable batch
+-- of _the_ candidate fragment, and then we choose the first one according to
+-- the ordering passed as argument.
+--
+-- PRECONDITION: The set of peers must be included in the peer order queue.
+selectThePeer ::
+  forall header block peer extra.
+  ( HasHeader header,
+    HeaderHash header ~ HeaderHash block,
+    Eq peer
+  ) =>
+  FetchDecisionPolicy header ->
+  (Point block -> Bool) ->
+  MaxSlotNo ->
+  PeersOrder peer ->
+  -- | The candidate fragment that we have selected to sync from, as suffix of
+  -- the immutable tip.
+  ChainSuffix header ->
+  -- | Association list of candidate fragments (as suffixes of the immutable
+  -- tip) and their associated peers.
+  [(ChainSuffix header, PeerInfo header peer extra)] ->
+  WithDeclined
+    (PeerInfo header peer extra)
+    (Maybe (ChainSuffix header, PeerInfo header peer extra))
+selectThePeer
+  FetchDecisionPolicy {blockFetchSize}
+  fetchedBlocks
+  fetchedMaxSlotNo
+  peersOrder
+  theCandidate
+  candidates = do
+    -- Filter out from the chosen candidate fragment the blocks that have
+    -- already been downloaded, but keep the blocks that have a request in
+    -- flight.
+    let fragments =
+          snd
+            <$> filterNotAlreadyFetched
+              fetchedBlocks
+              fetchedMaxSlotNo
+              theCandidate
+
+    -- Create a fetch request for the blocks in question The request is made
+    -- to fit in 1MB but ignores everything else. It is gross in that sense.
+    -- It will only be used to choose the peer to fetch from, but we will
+    -- later craft a more refined request for that peer.
+    let grossRequest =
+          selectBlocksUpToLimits
+            blockFetchSize
+            0 -- number of request in flight
+            maxBound -- maximum number of requests in flight
+            0 -- bytes in flight
+            (1024 * 1024) -- maximum bytes in flight; one megabyte
+            <$> fragments
+
+    -- For each peer, check whether its candidate contains the gross request
+    -- in its entirety, otherwise decline it.
```

Comment:
You're correct about the expectation that if the grossRequest can be served, then they will be able to serve the refined request. It does deserve to be set explicitly, though.

Reply:
Is 0be0889f6f0bc7687556f19edff9e05f4821d62f enough or would you want more?

Reply:
Hum, so the fact that the peer has the `grossRequest` means it can definitely create a request of at least one block (as long as the gross request is not empty; maybe I should add an assertion). And then this request can either be included in the `grossRequest` (for instance if the peer has tighter bytes-in-flight limits) but it can also include the `grossRequest` (if the peer has much higher bytes-in-flight limits and also has more blocks than this).

Reply:
It is very much worth documenting somewhere. I don't think this in-code documentation is the best place, though.

Reply:
WDYT of 04be7c4a8c7f9093db1e5514a2d83f6f2b8ceed5?